pkg/healthz/healthz.go

```diff
@@ -0,0 +1,132 @@
+	encryptCkFound := false
+	for _, ck := range healthzCheckers {
+		_, ok := checkers[ck]
+		if !ok {
+			return nil, fmt.Errorf("checker %q unknown", ck)
+		}
+		if ck == "encrypt" {
+			encryptCkFound = true
+			continue
+		}
+		if ck == "decrypt" && !encryptCkFound { // "encrypt" must precede
+			return nil, ErrHandler
+		}
+	}
+	return &handler{
+		mu:              sync.RWMutex{},
+		srv:             srv,
+		cli:             cli,
+		healthzCheckers: healthzCheckers,
+	}, nil
+}
+
+type handler struct {
+	mu              sync.RWMutex
+	srv             *plugin.Plugin
+	cli             pb.KeyManagementServiceClient
+	healthzCheckers []string
+	cipher          []byte
+}
+
+func (hd *handler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	res, err := plugin.Check(hd.cli)
+	if err != nil {
+		rw.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(rw, http.StatusText(http.StatusInternalServerError))
+		zap.L().Error("failed healthz version", zap.Error(err))
+		return
+	}
+	zap.L().Info("passed healthz version", zap.String("version", res))
+	if len(hd.healthzCheckers) == 0 {
+		rw.WriteHeader(http.StatusOK)
+		fmt.Fprint(rw, http.StatusText(http.StatusOK))
+		return
+	}
+
+	for _, name := range hd.healthzCheckers {
+		zap.L().Info("checking", zap.String("name", name))
+		checker, ok := checkers[name]
+		if !ok {
+			rw.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprint(rw, http.StatusText(http.StatusInternalServerError))
+			zap.L().Warn("checker not found", zap.String("name", name))
+			return
+		}
+		err := checker(hd)
+		if err != nil {
+			rw.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprint(rw, fmt.Errorf("failed healthz KMS API (%v)", err))
+			return
+		}
+	}
+	rw.WriteHeader(http.StatusOK)
+	fmt.Fprint(rw, http.StatusText(http.StatusOK))
+	zap.L().Info("passed healthz KMS API")
+}
+
+var checkers = map[string]func(hd *handler) error{
+	"encrypt":      checkerEncrypt,
+	"decrypt":      checkerDecrypt,
+	"kms-throttle": checkerKMSThrottle,
+}
+
+func checkerEncrypt(hd *handler) error {
+	resp, err := hd.srv.Encrypt(context.Background(), &pb.EncryptRequest{Plain: []byte("healthz-encrypt")})
+	if err != nil {
+		zap.L().Warn("failed healthz encrypt", zap.Error(err))
+		return err
+	}
+
+	// store cipher text with latest CMK
+	hd.mu.Lock()
+	hd.cipher = resp.Cipher
+	hd.mu.Unlock()
+
+	zap.L().Info("passed healthz encrypt", zap.Int("bytes", len(hd.cipher)))
+	return nil
+}
+
+func checkerDecrypt(hd *handler) error {
+	hd.mu.RLock()
+	cipher := hd.cipher
+	hd.mu.RUnlock()
+
+	resp, err := hd.srv.Decrypt(context.Background(), &pb.DecryptRequest{Cipher: cipher})
+	if err != nil {
+		zap.L().Warn("failed healthz decrypt", zap.Error(err))
+		return err
+	}
+	zap.L().Info("passed healthz decrypt", zap.Int("bytes", len(resp.Plain)))
+	return nil
+}
+
+func checkerKMSThrottle(hd *handler) error {
+	_, err := hd.srv.Encrypt(context.Background(), &pb.EncryptRequest{Plain: []byte("healthz-kms-throttle")})
```

Comment:
I think this can be folded into your `checkerEncrypt` since you can differentiate the throttle error from other errors.  No need to call `kms:Encrypt` twice per health check